src/pwm.rs

```diff
@@ -0,0 +1,206 @@
+use embedded_hal::Pwm;
+use gd32vf103_pac::{TIMER0, TIMER1, TIMER2, TIMER3, TIMER4};
+
+use crate::gpio::{Alternate, PushPull};
+use crate::gpio::gpioa::*;
+use crate::gpio::gpiob::*;
+use crate::rcu::{APB1, APB2, Clocks};
+use crate::time::{Hertz, U32Ext};
+
+pub trait PwmChannelPin<TIMER> {}
+macro_rules! pwm_pin {
+    ($timer:ty, $pin:ident) => {
+        impl PwmChannelPin<$timer> for $pin<Alternate<PushPull>> {}
+    }
+}
+
+// TODO: Handle alternate mappings.
+
+pwm_pin!(TIMER0, PA8);
+pwm_pin!(TIMER0, PA9);
+pwm_pin!(TIMER0, PA10);
+pwm_pin!(TIMER0, PA11);
+
+pwm_pin!(TIMER1, PA0);
+pwm_pin!(TIMER1, PA1);
+pwm_pin!(TIMER1, PA2);
+pwm_pin!(TIMER1, PA3);
+
+pwm_pin!(TIMER2, PA6);
+pwm_pin!(TIMER2, PA7);
+pwm_pin!(TIMER2, PB0);
+pwm_pin!(TIMER2, PB1);
+
+pwm_pin!(TIMER3, PB6);
+pwm_pin!(TIMER3, PB7);
+pwm_pin!(TIMER3, PB8);
+pwm_pin!(TIMER3, PB9);
+
+// Channel 3 only.
+// TODO: Enforce this in the code.
+pwm_pin!(TIMER4, PA3);
+
+pub struct PwmTimer<'a, TIMER> {
+    timer: TIMER,
+    clocks: Clocks,
+    max_duty_cycle: u16,
+    period: Hertz,
+    duty: [f32; 4],
+    ch0: Option<&'a dyn PwmChannelPin<TIMER>>,
+    ch1: Option<&'a dyn PwmChannelPin<TIMER>>,
+    ch2: Option<&'a dyn PwmChannelPin<TIMER>>,
+    ch3: Option<&'a dyn PwmChannelPin<TIMER>>,
+}
+
+#[derive(Copy, Clone, Debug)]
+pub struct Channel(pub u8);
+
+macro_rules! pwm_timer {
+    ($($TIM:ident: ($tim:ident, $timXen:ident, $timXrst:ident, $APB:ident, $cvtype:ty),)+) => {
+        $(
+            impl<'a> PwmTimer<'a, $TIM> {
+                pub fn new(timer: $TIM,
+                           clocks: Clocks,
+                           apb: &mut $APB,
+                           ch0: Option<&'a dyn PwmChannelPin<$TIM>>,
+                           ch1: Option<&'a dyn PwmChannelPin<$TIM>>,
+                           ch2: Option<&'a dyn PwmChannelPin<$TIM>>,
+                           ch3: Option<&'a dyn PwmChannelPin<$TIM>>) -> Self {
+                    let timer = PwmTimer {
+                        timer,
+                        clocks,
+                        max_duty_cycle: 0,
+                        period: 0.hz(),
+                        duty: [0f32; 4],
+                        ch0,
+                        ch1,
+                        ch2,
+                        ch3,
+                    };
+
+                    apb.en().modify(|_r, w| w.$timXen().set_bit());
+                    apb.rstr().modify(|_r, w| w.$timXrst().set_bit());
+                    apb.rstr().modify(|_r, w| w.$timXrst().clear_bit());
```

Comment:
It's better to use `$TIM::enable();` and `$TIM::reset();` here

Reply:
Thanks, I missed those traits. Any idea why they are not implemented for TIMER4-TIMER6? I can add that too, but I wanted to understand if there is some limitation about those timers.

Reply:
I think that I forgot to add those.